nav2_smac_planner/src/analytic_expansion.cpp

```diff
@@ -78,83 +86,61 @@ typename AnalyticExpansion<NodeT>::NodePtr AnalyticExpansion<NodeT>::tryAnalytic
     // Always run the expansion on the first run in case there is a
     // trivial path to be found
     if (analytic_iterations <= 0) {
-      // Reset the counter and try the analytic path expansion
       analytic_iterations = desired_iterations;
-      AnalyticExpansionNodes analytic_nodes =
-        getAnalyticPath(current_node, goal_node, getter, current_node->motion_table.state_space);
-      if (!analytic_nodes.empty()) {
-        // If we have a valid path, attempt to refine it
-        NodePtr node = current_node;
-        NodePtr test_node = current_node;
-        AnalyticExpansionNodes refined_analytic_nodes;
-        for (int i = 0; i < 8; i++) {
-          // Attempt to create better paths in 5 node increments, need to make sure
-          // they exist for each in order to do so (maximum of 40 points back).
-          if (test_node->parent && test_node->parent->parent && test_node->parent->parent->parent &&
-            test_node->parent->parent->parent->parent &&
-            test_node->parent->parent->parent->parent->parent)
-          {
-            test_node = test_node->parent->parent->parent->parent->parent;
-            refined_analytic_nodes =
-              getAnalyticPath(test_node, goal_node, getter, test_node->motion_table.state_space);
-            if (refined_analytic_nodes.empty()) {
-              break;
-            }
-            analytic_nodes = refined_analytic_nodes;
-            node = test_node;
-          } else {
-            break;
-          }
+      bool found_valid_expansion = false;
+      unsigned int number_of_checked_goal = 0;
+
+      // First check the coarse search resolution goals
+      while(number_of_checked_goal < coarse_search_goal_size) {
+        NodePtr current_goal_node = goals_to_expand[number_of_checked_goal];
+        AnalyticExpansionNodes analytic_nodes =
+          getAnalyticPath(
+          current_node, current_goal_node, getter,
+          current_node->motion_table.state_space);
+        if (!analytic_nodes.empty()) {
+          NodePtr node = current_node;
+          refineAnalyticPath(
+            current_goal_node, getter, node, analytic_nodes, best_score);
+          // Update the best score if we found a better path
+          current_best_analytic_nodes = analytic_nodes;
+          current_best_goal = current_goal_node;
+          current_best_node = node;
+          found_valid_expansion = true;
+          break;
```

Comment:
Note: the break here then wouldn't attempt all coarse resolutions, it ends early. We'd need to iterate through the full set so we actually get the "best" one -- sine in the 'fine search' we don't reattempt any of the coarse resolutions. They're simply skipped and could actually be better than the first one found.

Also, I think this should be updated to store the current best if its better than the last score. `refineAnalyticPath` should be responsible for attempting to refine the path and then returning the best score and its nodes only. Instead, it should have a return of `best_score` that we compare to our current best score and only update `current_best_*` if its an improvement. We do that in the fine search but not sure why we don't do that here as well.

Reply:
The idea here was to search the coarse resolution and as soon as we find a goal, we do the whole search. We loop through the coarse resolution and if don't find any goal, we don't do a fine search. First we search with a while loop until the numbers of coarse search goals, if we don't find any solutions, we never set the goal found to true so everything is ok. When we find a goal, we break, but we still search the remaining coarse resolution, in the code line "for (size_t i = number_of_checked_goal; i < goals_to_expand.size(); i++) {" after we find a valid goal, we start the for loop from the last checked goal. 

This would be clearer with the new modification we discussed earlier with two vectors.